webapp/src/shared/components/Input/SearchForm.tsx

```diff
@@ -0,0 +1,36 @@
+import { useEffect, useState } from 'react';
+import { Input, InputVariant, RowsListProps } from '..';
+import { IconVariant } from '../Icon/Icon';
+import { getRows } from '../../../pages/Users/Users';
+import './SearchForm.css';
+
+type SearchProps = {
+  url: string;
+  setChange: React.Dispatch<React.SetStateAction<RowsListProps>>;
+};
+
+export default function SearchForm({ url, setChange }: SearchProps) {
+  const [search, setSearch] = useState('');
+  useEffect(() => {
+    const fetchRowsList = async () => {
+      const lRowsList = await getRows(`${url}?search=${search}`);
```

Comment:
If we really want to make it generic, then we need to pass getRows as a callBack in a prop as well, otherwise it's still specific to user Search. GetRows is specific to users.

Reply:
Yes, your are right. Changed

Reply:
We could do something like this:

<details>

```diff
diff --git a/webapp/src/pages/Users/Users.tsx b/webapp/src/pages/Users/Users.tsx
index 2e2c736..52a8278 100644
--- a/webapp/src/pages/Users/Users.tsx
+++ b/webapp/src/pages/Users/Users.tsx
@@ -2,12 +2,12 @@ import './Users.css';
 import {
   IconVariant,
   NavigationBar,
+  RowItem,
   RowsList,
-  RowsListProps,
   SearchForm,
   SmallAvatar,
 } from '../../shared/components';
-import { useEffect, useState } from 'react';
+import { useState } from 'react';
 import {
   USER_URL,
   USERS_EP_URL,
@@ -24,31 +24,26 @@ interface User {
   readonly createdAt: Date;
 }
 
-export const getRows = async (url: string): Promise<RowsListProps> => {
-  const users = (await fetch(url).then((response) => response.json())) ?? [];
-  return {
-    rows: users.map((user: User) => {
-      return {
-        iconVariant: IconVariant.ARROW_FORWARD,
-        avatarProps: {
-          url: user.avatarId
-            ? `${USERS_EP_URL}/${user.id}/avatar`
-            : WILDCARD_AVATAR_URL,
-          status: 'offline',
-        },
-        url: `${USERS_URL}/${user.username}`,
-        title: user.username,
-        subtitle: 'level x',
-        key: user.id,
-      };
-    }),
-  };
+const mapUsersToRows = (users: User[]): RowItem[] => {
+  return users.map((user) => {
+    return {
+      iconVariant: IconVariant.ARROW_FORWARD,
+      avatarProps: {
+        url: user.avatarId
+          ? `${USERS_EP_URL}/${user.id}/avatar`
+          : WILDCARD_AVATAR_URL,
+        status: 'offline',
+      },
+      url: `${USERS_URL}/${user.username}`,
+      title: user.username,
+      subtitle: 'level x',
+      key: user.id,
+    };
+  });
 };
 
 export function Users() {
-  const [usersList, setUsersList] = useState<RowsListProps>({ rows: [] });
-
-  useEffect(() => {}, [usersList]);
+  const [users, setUsers] = useState<User[]>([]);
 
   return (
     <div className="users">
@@ -58,10 +53,10 @@ export function Users() {
         </Link>
       </div>
       <div className="users-search">
-        <SearchForm url={`${USERS_EP_URL}`} setChange={setUsersList} />
+        <SearchForm url={`${USERS_EP_URL}?search=`} setChange={setUsers} />
       </div>
       <div className="users-rows">
-        {usersList.rows && <RowsList rows={usersList.rows} />}
+        <RowsList rows={mapUsersToRows(users)} />
       </div>
       <div className="users-navigation">
         <NavigationBar />
diff --git a/webapp/src/shared/components/Input/SearchForm.tsx b/webapp/src/shared/components/Input/SearchForm.tsx
index ae07186..1bd783f 100644
--- a/webapp/src/shared/components/Input/SearchForm.tsx
+++ b/webapp/src/shared/components/Input/SearchForm.tsx
@@ -1,24 +1,19 @@
 import { useEffect, useState } from 'react';
-import { Input, InputVariant, RowsListProps } from '..';
+import { Input, InputVariant } from '..';
 import { IconVariant } from '../Icon/Icon';
-import { getRows } from '../../../pages/Users/Users';
 import './SearchForm.css';
 
-type SearchProps = {
+type SearchProps<T> = {
   url: string;
-  setChange: React.Dispatch<React.SetStateAction<RowsListProps>>;
+  setChange: React.Dispatch<React.SetStateAction<T>>;
 };
 
-export default function SearchForm({ url, setChange }: SearchProps) {
+export default function SearchForm<T>({ url, setChange }: SearchProps<T>) {
   const [search, setSearch] = useState('');
-  useEffect(() => {
-    const fetchRowsList = async () => {
-      const lRowsList = await getRows(`${url}?search=${search}`);
-      setChange({ ...lRowsList });
-    };
-    fetchRowsList().catch((e) => console.error(e));
-  }, [search, setChange, url]);
-
+  const results: T | null = useData(`${url}${search}`);
+  if (results) {
+    setChange(results);
+  }
   return (
     <div className="search-form">
       <Input
@@ -34,3 +29,21 @@ export default function SearchForm({ url, setChange }: SearchProps) {
     </div>
   );
 }
+
+function useData<T>(url: string): T | null {
+  const [data, setData] = useState<T | null>(null);
+  useEffect(() => {
+    let ignore = false;
+    fetch(url)
+      .then((response) => response.json())
+      .then((json) => {
+        if (!ignore) {
+          setData(json);
+        }
+      });
+    return () => {
+      ignore = true;
+    };
+  }, [url]);
+  return data;
+}

```

</details>

The `useData` hook is from the beta [docs](https://beta.reactjs.org/learn/you-might-not-need-an-effect#fetching-data). I've only added typing and generics.

Reply:
[Here](https://github.com/42AGV/ft_transcendence/pull/129/files#diff-1a55e244fd286a4238b878c647d9feff58df84d40ffb8638930b75143f86cf50) , @gbudau

Reply:
I've just realized that with these solutions, trying to make the search form more generic, if you open the Network Monitor in the browser it is continuously sending request.
This didn't happened with solution of commit [split into searchUserForm](https://github.com/42AGV/ft_transcendence/pull/118/commits/a667a738f3765dcecf0e177e392ffcd2036f8cad) [a667a73](https://github.com/42AGV/ft_transcendence/pull/118/commits/a667a738f3765dcecf0e177e392ffcd2036f8cad)
Please check if you have the same issue

Reply:
It doesn't happen anymore. I don't know why but it seems to works properly. Sorry

Reply:
I still haven't solved compatibility with custom hook from #129 user page PR

Reply:
Thank you :+1: 

There is a console error because we call `setChange` before this component gets render(which in turn re-renders the `Users` component):

```
Warning: Cannot update a component (`Users`) while rendering a different component (`SearchForm`). To locate the bad setState() call inside `SearchForm`, follow the stack trace as described in https://reactjs.org/link/setstate-in-render
```

We could set the result in an `useEffect`

```
  useEffect(() => {
    if (result) {
      setChange(result);
    }
  }, [result, setChange]);
```

Reply:
Also, I just pulled and I can still see this warning in the console
![Screenshot from 2022-08-15 15-49-41](https://user-images.githubusercontent.com/2774420/184638004-a5fdaa22-4ee9-4da5-b3fc-4297593688f1.png)

Reply:
You are right. Done